docs/internal/types.md

```diff
@@ -0,0 +1,359 @@
+  increment();
+};
+
+// This class inherits, so it gets the field `common` from the interface
+class MyType : Example
+{
+  let field : I32;
+
+  // Some other method
+  innerSum(x : I32) : I32
+  {
+    let res : I32 = x + field;
+    res
+  }
+
+  // Implements `increment`
+  increment()
+  {
+    self.common += 1;
+  }
+};
+
+// This class implements the interface, so it must provide the `common` field
+class OtherType {
+  // Must provide the field
+  var common : I64;
+
+  // Implements `increment`
+  increment()
+  {
+    self.common += 2;
+  }
+};
+```
+
+### Fields
+
+Define objects that belong to the class or interface and will be instantiated (ie. use memory) on every instance of the class.
+
+Fields can be initialised in two ways:
+ * `var`: Must be initialised on construction, but can change after.
+ * `let`: Must be initialised on construction, and cannot change later, like const fields in C++.
+
+They can also have a keyword `embed`, which means they're placed in the struct directly, as opposed to the default of being a pointer to the data.
+
+For example
+```ts
+// This struct layout would be { F64, F64 }
+class Point
+{
+  embed var x : F64;
+  embed var y : F64;
+};
+
+// This struct layout would be { Ref[Point], Ref[Point] }, same as { USize, USize }
+// With `Ref` being the notion of the "address of" an object.
+class Line
+{
+  var orig : Point;
+  var dest : Point;
+};
+```
+
+### Methods
+
+Methods are functions that belong to a class or interface which can have arguments and a return type.
+
+Methods can be called statically (via `Class::method()`) if the compiler can infer the concrete type, or dynamically (via `object.method()`) otherwise.
+Methods that change the fields in the class must receive a `self : Self` argument and use as the field accessor (ex. `self.field`).
+
+In the following example, all calls are statically defined:
+```ts
+class Foo
+{
+  method1() { ... }
+  method2(self : Self) { ... }
+};
+
+...
+
+// Class syntax, static call
+Foo::method1();
+
+// Object syntax, static call
+var x : Foo = ...;
+x.method1(); // Calls Foo::method1()
+x.method2(); // Calls Foo::method2(x)
+```
+
+In the next example, interfaces may stop the compiler from finding the concrete type:
+```ts
+interface Bar
+{
+  method1() { ... }
+  method2(self : Self) { ... }
+}
+
+...
+
+// Object syntax, dynamic call
+var y : Bar = ...;
+y.method1(); // Could be Foo, but not sure, so use `y`'s vtable.
+y.method2(); // Same here, but Bar::method2(y);
+```
+
+The compiler will make sure that all classes that implement an interface will have the same methods to be deterministically positioned so that `Interface::method()` doesn't end up with multiple levels of indirection.
+
+### Constructors
+
+Verona's built-in `new` function fully constructs an object and so must be called with values for all of the fields in the object.
+It may be used only from static methods on the class and so is usually used by factory methods.
```

Comment:
What does "static" mean here?  I don't think you mean to preclude, for example,

```ts
class Mushroom
{
  let genus : String;

  clone(self: Mushroom) : Mushroom
  {
    return new Mushroom(self.genus);
  }

  create(genus : String) : Mushroom
  {
    return new Mushroom(genus);
  }
}

...
let c1 = Mushroom::clone(original);
let c2 = c1.clone();
...
```

Reply:
Static here is used to mean methods that don't have a self parameter (i.e. class methods), but the doc doesn't mention what a self parameter is.  Actually, I believe there's no such restriction on `new` necessary: allowing an instance method to construct more instances directly isn't a problem (as you point out, for `clone`), but at least one class method must call `new` if you ever want to have an instance that reaches this point.

I'm not quite sure how `clone` interacts with `embed`.

Reply:
I agree it's confusing. I don't know of a better name here. I'd be happy to change the nomenclature to whatever is more appropriate and less overloaded.

Reply:
I've added a short comment on the `self` argument to help, and with the new section on static and dynamic methods, it should be fine.